order/src/main/java/br/com/estudo/storechallenge/order/repository/OrderRepository.java

```diff
@@ -16,14 +15,14 @@
 
     @Query("SELECT SUM(oi.quantity * oi.unitPrice) " +
             " FROM OrderItem oi " +
-            " WHERE TO_CHAR(oi.order.creationDate, 'MM/YYYY') = :month " +
+            " WHERE date_trunc('day', oi.order.creationDate) BETWEEN :startDate AND :endDate " +
             "  AND oi.order.storeId = :storeId")
-    Optional<BigDecimal> sumTotalSalesByStoreIdAndMonth(Long storeId, String month);
+    Optional<BigDecimal> sumTotalSalesByStoreIdAndIntervalDates(Long storeId, Date startDate, Date endDate);
```

Comment:
Porque usar Date e não localDate?

Reply:
Refiz alguns testes e tentei algumas outras opções (_todos com parâmetro em LocalDate_):
**1. Comparação direto com o parâmetro**
`WHERE date_trunc('day', oi.order.creationDate) = :day`
Parameter value [2020-01-03] did not match expected type [java.time.LocalDateTime (n/a)];

**2. Parâmetro em LocalDateTime**
InvalidDataAccessApiUsageException: Parameter value [2020-01-03T00:00] did not match expected type [java.util.Date (n/a)]; nested exception is java.lang.IllegalArgumentException: Parameter value [2020-01-03T00:00] did not match expected type [java.util.Date (n/a)]

**3. Comparação entre date_trunc com date_trunc**
`WHERE date_trunc('day', oi.order.creationDate) = date_trunc('day', :day)`
ERROR: function date_trunc(unknown, unknown) is not unique
InvalidDataAccessResourceUsageException: could not extract ResultSet; SQL [n/a]; nested exception is org.hibernate.exception.SQLGrammarException: could not extract ResultSet

**4. Conversores**
`LocalDateTimeAttributeConverter implements AttributeConverter<LocalDateTime, Timestamp>`
LocalDateTimeAttributeConverter implements AttributeConverter<LocalDateTime, Timestamp>

Reply:
Realmente não sei mais o que buscar Uri... :(

Reply:
Tenta o seguinte no service:
`    public StoreTotalSalesResponse findTotalSalesByStoreId(Long storeId) {
        log.info("Calculating the total sales to store: {}", storeId);

        LocalDate today = LocalDate.now();
        LocalDateTime startDate = today.withDayOfMonth(1).atStartOfDay();
        LocalDateTime endDate = today.with(TemporalAdjusters.lastDayOfMonth()).atTime(LocalTime.MAX);

        StoreTotalSalesResponse storeTotalSalesResponse = StoreTotalSalesResponse.builder()
                .storeId(storeId)
                .totalDay(orderRepository.sumTotalSalesByStoreIdAndIntervalDates(storeId, 
                        today.atStartOfDay(), today.atTime(LocalTime.MAX))
                        .orElse(new BigDecimal("0")))
                .totalMonth(orderRepository.sumTotalSalesByStoreIdAndIntervalDates(storeId,
                            startDate, endDate)
                        .orElse(new BigDecimal("0")))
                .build();

        return storeTotalSalesResponse;
    }
`
No repository, só terá um método:
`    @Query("SELECT SUM(oi.quantity * oi.unitPrice) " +
            " FROM OrderItem oi " +
            " WHERE oi.order.creationDate BETWEEN :startDate AND :endDate " +
            "  AND oi.order.storeId = :storeId")
    Optional<BigDecimal> sumTotalSalesByStoreIdAndIntervalDates(Long storeId,
                                                                LocalDateTime startDate,
                                                                LocalDateTime endDate);
`
Gostaria que me fale porque desta forma funciona e porque da outra não, é importante entender os motivos. Também que desvantagens vê nesta abordagem.

Reply:
Realmente funcionou.
Entendi que eu estava forçando a utilização do `date_trunc` na query e isso fazia com que o parâmetro de comparação fosse um `java.util.Date`. Porém não precisa ser assim. Como na Entity de Order, o atributo `creationDate` é do tipo LocalDateTime então faz mais sentido usar os parâmetros de intervalo como LocalDateTime e ao invés de remover a hora, criar um o limite mínimo (00:00:00..) no `startDate` e o limite máximo (23:59:59..) no `endDate` no Java utilizando as classes do java.time: `LocalTime` e `TemporalAdjusters`.
Tem algo mais que eu deixei passar?

Reply:
Não tenho certeza, mas acredito que o trunc transforma para sql.Date, e não existe transformação de LocalDate para sql.Date. 
Desta forma eliminamos um método, mas a dúvida é aquela mesma do começo. Qual é mais performática? ou, melhor, a performance pode ser um problema? Por que se pensamos numa tabela na ordem de 1.000 ou 10.000 registros, com certeza não precisamos nos preocupar com isso, mas se pensamos em tabelas na casa dos milhões de registros e consultas síncronas, talvez precisamos nos questionar isso

Reply:
Pensando aqui, eu acredito que a consulta restringindo uma data com o "igual", o banco consegue criar um index e agilizar a resposta, mas de qualquer forma precisaríamos de uma busca no mês (por intervalo).
O que poderíamos fazer é restringir que a busca não passe de uma quantidade de dias, por exemplo, no máximo 60 dias de intervalo.
Mas lembrando aqui, foi neste ponto que você comentou para fazermos primeiramente de forma síncrona e depois daríamos um jeito de mudar para assíncrona, não foi?

Reply:
Sim, deixa assim por ora. As vezes a optimização prematura pode ser mais dor de cabeça que solução